File: src/csharp/37 - Sudoku Solver.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CSharpConsoleApp
{
    class Program
    {
        static bool IsValidSudoku(char[][] board)
        {
            char[] cache = new char[9];

            bool IsUnique(char[] chars)
            {
                var set = new HashSet<char>(9);
                foreach (var c in chars)
                {
                    if (c != '.')
                    {
                        if (set.Contains(c))
                        {
                            return false;
                        }
                        else
                        {
                            set.Add(c);
                        }
                    }
                }
                return true;
            }
            bool CheckRow(int row)
            {
                return IsUnique(board[row]);
            }
            bool CheckColumn(int column)
            {
                for (int row = 0; row < 9; row++)
                {
                    cache[row] = board[row][column];
                }
                return IsUnique(cache);
            }
            bool CheckBox(int row, int column)
            {
                var index = 0;
                for (int i = 0; i < 3; i++)
                {
                    for (int j = 0; j < 3; j++)
                    {
                        cache[index++] = board[row + i][column + j];
                    }
                }
                return IsUnique(cache);
            }

            for (int i = 0; i < 9; i++)
            {
                if (!CheckRow(i) || !CheckColumn(i))
                {
                    return false;
                }
            }
            for (int i = 0; i < 9; i += 3)
            {
                for (int j = 0; j < 9; j += 3)
                {
                    if (!CheckBox(i, j))
                    {
                        return false;
                    }
                }
            }
            return true;
        }

        static void SolveSudoku(char[][] board)
        {
            HashSet<char> Candidates(int row, int column)
            {
                if (board[row][column] != '.')
                {
                    throw new ArgumentException($"position ({row},{column}) not empty.");
                    //return new List<char>() { board[row][column] };
                }
                var set = new HashSet<char>() { '1', '2', '3', '4', '5', '6', '7', '8', '9' };
                // remove item in row
                for (int i = 0; i < 9; i++)
                {
                    set.Remove(board[row][i]);
                }
                // remove item in column
                for (int i = 0; i < 9; i++)
                {
                    set.Remove(board[i][column]);
                }
                // remove item in box
                var (top, left) = (row / 3 * 3, column / 3 * 3);
                for (int i = 0; i < 3; i++)
                {
                    for (int j = 0; j < 3; j++)
                    {
                        set.Remove(board[top + i][left + j]);
                    }
                }
                return set;
            }

            bool FillFixed()
            {
                var changed = false;
                for (int i = 0; i < 9; i++)
                {
                    for (int j = 0; j < 9; j++)
                    {
                        if (board[i][j] == '.')
                        {
                            var candidates = Candidates(i, j);
                            if (candidates.Count == 1)
                            {
                                board[i][j] = candidates.First();
                                changed = true;
                            }
                        }
                    }
                }
                return changed;
            }

            bool Solve()
            {
                var (row, column) = (-1, -1);
                for (int i = 0; i < 9; i++)
                {
                    for (int j = 0; j < 9; j++)
                    {
                        if (board[i][j] == '.')
                        {
                            (row, column) = (i, j);
                        }
                    }
                }
                if (row == -1)
                {
                    return true;
                }
                var candidates = Candidates(row, column);
                if (candidates.Count == 0)
                {
                    return false;
                }
                foreach (var c in candidates)
                {
                    board[row][column] = c;
                    if (Solve())
                    {
                        return true;
                    }
                }
                board[row][column] = '.';
                return false;
            }

            while (FillFixed()) { }

            Solve();

        }

        static void PrettyPrint(char[][] board)
        {
            var str = board.Select(row => string.Join(',', row))
                .Aggregate((r1, r2) => r1 + "\n" + r2);
            Console.WriteLine(str);
        }

        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            var board = new char[][]
            {
                new char[] {'5','3','.','.','7','.','.','.','.'},
                new char[] {'6','.','.','1','9','5','.','.','.'},
                new char[] {'.','9','8','.','.','.','.','6','.'},
                new char[] {'8','.','.','.','6','.','.','.','3'},
                new char[] {'4','.','.','8','.','3','.','.','1'},
                new char[] {'7','.','.','.','2','.','.','.','6'},
                new char[] {'.','6','.','.','.','.','2','8','.'},
                new char[] {'.','.','.','4','1','9','.','.','5'},
                new char[] {'.','.','.','.','8','.','.','7','9'}
            };
            PrettyPrint(board);
            SolveSudoku(board);
            Console.WriteLine("============");
            PrettyPrint(board);
        }
    }
}

File: src/fsharp/37 - Sudoku Solver.fs

open System

// more F#-smell
let solveSudoku (board: char array array) = 
    // 将二维数组转换成坐标及值的序列
    let seq = seq {
        let mutable row = 0
        while row < 9 do
            let mutable column = 0
            while column < 9 do
                yield ((row, column), board.[row].[column])
                column <- column + 1
            row <- row + 1
    }

    // 找出当前空白位置的所有可行候选者
    let candidates row column = 
        if board.[row].[column] <> '.' then ArgumentException($"position ({row},{column}) not empty.") |> raise
        let (top, left) = (row/3, column/3)
        let pred (r, c) = r = row || c = column || (r / 3 = top && c / 3 = left)
        let total = ['1';'2';'3';'4';'5';'6';'7';'8';'9'] |> Set.ofList
        let other = seq |> Seq.filter (fst >> pred) |> Seq.map snd |> Set.ofSeq
        total - other

    // 填上所有“单候选”空白位置
    // 某空白位置被填入后，会立即影响其它空白位置的取值，因此必须循环执行、最终到达一个不动点
    let fillFixed () =
        let isSingle set = set |> Set.count = 1
        let head set = set |> Set.toSeq |> Seq.head
        let position = seq |> Seq.tryFind (fun ((row, column), char) -> char = '.' && candidates row column |> isSingle) |> Option.map fst
        match position with
        | Some (row, column) -> board.[row].[column] <- candidates row column |> head; true
        | None -> false

    // 递归对“多候选”空白位置进行尝试、最终命中结果
    let rec solve () =
        let position = seq |> Seq.tryFind (snd >> fun char -> char = '.') |> Option.map fst
        match position with
        | Some (row, column) ->
            let correct =
                candidates row column
                |> Seq.filter (fun c -> board.[row].[column] <- c; solve())
                |> Seq.tryHead
            match correct with
            | Some c -> board.[row].[column] <- c; true
            | None -> board.[row].[column] <- '.'; false
        | _ -> false

    while fillFixed () do ()
    solve () |> ignore

// just convert C# code
let solveSudoku (board: char array array): unit = 
    // 找出当前空白位置的所有可行候选者
    let candidates row column = 
        if board.[row].[column] <> '.' then ArgumentException($"position ({row},{column}) not empty.") |> raise
        let mutable set = ['1';'2';'3';'4';'5';'6';'7';'8';'9'] |> Set.ofList
        for i = 0 to 8 do
            set <- set.Remove board.[row].[i]
            set <- set.Remove board.[i].[column]
        let (top, left) = (row / 3 * 3, column / 3 * 3)
        for i = 0 to 2 do
            for j = 0 to 2 do 
                set <- set.Remove board.[top+i].[left+j]
        set
    // 填上所有“单候选”空白位置
    // 某空白位置被填入后，会立即影响其它空白位置的取值，因此必须循环执行、最终到达一个不动点
    let fillFixed () =
        let mutable changed = false
        for i = 0 to 8 do
            for j = 0 to 8 do
                if board.[i].[j] = '.' then
                    let candidates = candidates i j
                    if candidates.Count = 1 then
                        board.[i].[j] <- candidates |> Set.toList |> List.head
                        changed <- true
        changed
    // 递归对“多候选”空白位置进行尝试、最终命中结果
    let rec solve () =
        let mutable row = -1
        let mutable column = -1
        for i = 0 to 8 do
            for j = 0 to 8 do
                if board.[i].[j] = '.' then
                    row <- i
                    column <- j
        if row <> -1 then
            let candidates = candidates row column
            if candidates.IsEmpty |> not then 
                let mutable succ = false
                for c in candidates do
                    if not succ then board.[row].[column] <- c
                    if not succ && solve() then succ <- true
                if not succ then board.[row].[column] <- '.'
                succ
            else false
        else false
    while fillFixed() do 
    solve() |> ignore


[<EntryPoint>]
let main argv =
    printfn "Hello World from F#!"

    let board = [|
        [|'5';'3';'.';'.';'7';'.';'.';'.';'.'|];
        [|'6';'.';'.';'1';'9';'5';'.';'.';'.'|];
        [|'.';'9';'8';'.';'.';'.';'.';'6';'.'|];
        [|'8';'.';'.';'.';'6';'.';'.';'.';'3'|];
        [|'4';'.';'.';'8';'.';'3';'.';'.';'1'|];
        [|'7';'.';'.';'.';'2';'.';'.';'.';'6'|];
        [|'.';'6';'.';'.';'.';'.';'2';'8';'.'|];
        [|'.';'.';'.';'4';'1';'9';'.';'.';'5'|];
        [|'.';'.';'.';'.';'8';'.';'.';'7';'9'|]
    |]
    printfn "%A" board
    solveSudoku board
    printfn "%A" board

    0